File: app/securesocial/provider/providers/FoursquareProvider.java
/**
 * Copyright 2011 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.provider.providers;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import play.Logger;
import play.libs.WS;
import securesocial.provider.*;

import java.util.Map;

/**
 * A provider for Foursquare 
 */
public class FoursquareProvider extends OAuth2Provider {
    private static final String SELF_API = "https://api.foursquare.com/v2/users/self?oauth_token=%s";
    private static final String META = "meta";
    private static final String CODE = "code";
    private static final String ERROR_TYPE = "errorType";
    private static final String ERROR_DETAIL = "errorDetail";
    private static final String RESPONSE = "response";
    private static final String USER = "user";
    private static final String ID = "id";
    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";
    private static final String SPACE = " ";
    private static final String PHOTO = "photo";
    private static final String CONTACT = "contact";
    private static final String EMAIL = "email";

    public FoursquareProvider() {
        super(ProviderType.foursquare);
    }

    @Override
    protected void fillProfile(SocialUser user, Map<String, Object> authContext) {
        JsonObject me = WS.url(SELF_API, user.accessToken).get().getJson().getAsJsonObject();
        JsonObject meta = me.getAsJsonObject(META);

        if ( meta.get(CODE).getAsInt() != 200 ) {
            final String errorType = meta.get(ERROR_TYPE).getAsString();
            final String errorDetail = meta.get(ERROR_DETAIL).getAsString();
            Logger.error("Error retrieving profile information from Foursquare. Error type: %s, detail: %s.", errorType, errorDetail);
            throw new AuthenticationException();
        }

        JsonObject response = me.getAsJsonObject(RESPONSE);
        if ( response == null ) {
            throw new AuthenticationException();
        }

        JsonObject userInfo = response.getAsJsonObject(USER);
        if( userInfo == null ) {
            throw new AuthenticationException();
        }

        user.id.id = userInfo.get(ID).getAsString();
        user.displayName = userInfo.get(FIRST_NAME).getAsString();
        final JsonElement lastName = userInfo.get(LAST_NAME);
        if ( lastName != null ) {
            user.displayName = fullName(user.displayName, lastName.getAsString());
        }
        user.avatarUrl = userInfo.get(PHOTO).getAsString();
        final JsonObject contact = userInfo.getAsJsonObject(CONTACT);
        if ( contact != null ) {
            final JsonElement userEmail = contact.get(EMAIL);
            if ( userEmail != null ) {
                user.email = userEmail.getAsString();
            }
        }
    }

    static String fullName(String first, String last) {
        return new StringBuilder(first).append(SPACE).append(last).toString();
    }
}

File: module-code/app/securesocial/core/providers/FoursquareProvider.scala
/**
 * Copyright 2013 Brian Porter (poornerd at gmail dot com) - twitter: @poornerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.core.providers

import securesocial.core._
import securesocial.core.services.{CacheService, RoutesService}

import scala.concurrent.Future

/**
 * A Foursquare provider
 *
 */
class FoursquareProvider(routesService: RoutesService,
                         cacheService: CacheService,
                         client: OAuth2Client)
  extends OAuth2Provider(routesService, client, cacheService)
{
  val GetAuthenticatedUser = "https://api.foursquare.com/v2/users/self?v=20140404oauth_token=%s"
  val AccessToken = "access_token"
  val TokenType = "token_type"
  val Message = "message"
  val Id = "id"
  val Response = "response"
  val User = "user"
  val Contact = "contact"
  val LastName = "lastName"
  val FirstName = "firstName"
  val AvatarUrl = "photo"
  val Email = "email"
  val Prefix = "prefix"
  val Suffix = "suffix"

  override val id = FoursquareProvider.Foursquare

  def fillProfile(info: OAuth2Info): Future[BasicProfile] = {
    import scala.concurrent.ExecutionContext.Implicits.global
    client.retrieveProfile(GetAuthenticatedUser.format(info.accessToken)).map { me =>
        (me \ "response" \ "user").asOpt[String] match {
          case Some(msg) =>
            logger.error("[securesocial] error retrieving profile information from Foursquare. Message = %s".format(msg))
            throw new AuthenticationException()
          case _ =>
            val userId = (me \ Response \ User \ Id).as[String]
            val lastName = (me \ Response \ User \ LastName).asOpt[String]
            val firstName = (me \ Response \ User \ FirstName).asOpt[String]
            val avatarUrlPart1 = (me \ Response \ User \ AvatarUrl \ Prefix).asOpt[String]
            val avatarUrlPart2 = (me \ Response \ User \ AvatarUrl \ Suffix).asOpt[String]
            val avatarUrl = for (prefix <- avatarUrlPart1; postfix <- avatarUrlPart2) yield prefix + "100x100" + postfix
            val email = (me \ Response \ User \ Contact \ Email).asOpt[String].filter(!_.isEmpty)
            BasicProfile(id, userId, firstName, lastName, None, email, avatarUrl, authMethod, oAuth2Info = Some(info))
        }
    } recover {
      case e: AuthenticationException => throw e
      case e  =>
        logger.error( "[securesocial] error retrieving profile information from Foursquare", e)
        throw new AuthenticationException()
    }
  }
}

object FoursquareProvider {
  val Foursquare = "foursquare"
}
